velox/functions/sparksql/Mask.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utility>
+
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/expression/VectorWriters.h"
+#include "velox/functions/lib/Re2Functions.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+// mask(string) -> string
+// mask(string, upperChar) -> string
+// mask(string, upperChar, lowerChar) -> string
+// mask(string, upperChar, lowerChar, digitChar) -> string
+// mask(string, upperChar, lowerChar, digitChar, otherChar) -> string
+//
+// Masks the characters of the given string value with the provided specific
+// characters respectively. Upper-case characters are replaced with the second
+// argument. Default value is 'X'. Lower-case characters are replaced with the
+// third argument. Default value is 'x'. Digit characters are replaced with the
+// fourth argument. Default value is 'n'. Other characters are replaced with the
+// last argument. Default value is NULL and the original character is retained.
+// If the provided nth argument is NULL, the related original character is
+// retained.
+class MaskFunction final : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    // Get the decoded vectors out of arguments.
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* strings = decodedArgs.at(0);
+    DecodedVector* upperChars = args.size() >= 2 ? decodedArgs.at(1) : nullptr;
+    DecodedVector* lowerChars = args.size() >= 3 ? decodedArgs.at(2) : nullptr;
+    DecodedVector* digitChars = args.size() >= 4 ? decodedArgs.at(3) : nullptr;
+    DecodedVector* otherChars = args.size() >= 5 ? decodedArgs.at(4) : nullptr;
+
+    BaseVector::ensureWritable(rows, VARCHAR(), context.pool(), result);
+    auto* flatResult = result->as<FlatVector<StringView>>();
+    auto getMaskedChar = [&](DecodedVector* inputChars,
+                             const VectorPtr& arg,
+                             int index,
+                             std::optional<char> maskedChar,
+                             const char* charType) -> std::optional<char> {
```

Comment:
this is align with the message of what vinalila spark showing when throw the exception,  I'd like to keep it, otherwise the user may miss which type replacing char not satifity the requirement.